chips/sam4l/src/eic.rs

```diff
@@ -0,0 +1,375 @@
+// ISR: 0: An interrupt event has not occurred.
+//      1: An interrupt event has occurred.
+// ICR: Writing a one to this bit will clear the corresponding bit in ISR.
+// MODE:    0: The external interrupt is edge triggered.
+//          1: The external interrupt is level triggered.'
+// EDGE:    0: The external interrupt triggers on falling edge.
+//          1: The external interrupt triggers on rising edge.
+// LEVEL:   0: The external interrupt triggers on low level.
+//          1: The external interrupt triggers on high level.
+// FILTER:  0: The external interrupt is not filtered.
+//          1: The external interrupt is filtered.
+// ASYNC:   0: The external interrupt is synchronized to CLK_SYNC.
+//          1: The external interrupt is asynchronous.
+// EN: Writing a one to this bit will enable the corresponding external interrupt.
+// DIS: Writing a one to this bit will disable the corresponding external interrupt.
+// CTRL:    0: The corresponding external interrupt is disabled.
+//          1: The corresponding external interrupt is enabled.
+
+register_bitfields![
+    u32,
+    Interrupt [
+        INT OFFSET(0) NUMBITS(32) []
+    ],
+    Test [
+        //0: This bit disables external interrupt test mode.
+        //1: This bit enables external interrupt test mode.
+        TESTEN 31,
+
+        // Writing a zero to this bit will set the input value to INTn to zero, if test mode is enabled. 
+        // Writing a one to this bit will set the input value to INTn to one, if test mode is enabled.
+        INT30 30,
+        INT29 29,
+        INT28 28,
+        INT27 27,
+        INT26 26,
+        INT25 25,
+        INT24 24,
+        INT23 23,
+        INT22 22,
+        INT21 21,
+        INT20 20,
+        INT19 19,
+        INT18 18,
+        INT17 17,
+        INT16 16,
+        INT15 15,
+        INT14 14,
+        INT13 13,
+        INT12 12,
+        INT11 11,
+        INT10 10,
+        INT9 9,
+        INT8 8,
+        INT7 7,
+        INT6 6,
+        INT5 5,
+        INT4 4,
+        INT3 3,
+        INT2 2,
+        INT1 1,
+        NMI 0   // Non-Maskable Interrupt
+    ]
+];
+
+// Page 59 of SAM4L data sheet
+const EIC_BASE: StaticRef<EicRegisters> =
+    unsafe { StaticRef::new(0x400F1000 as *const EicRegisters) };
+
+pub struct Eic<'a> {
+    registers: StaticRef<EicRegisters>,
+    enabled: Cell<bool>,
+    callbacks: [OptionalCell<&'a hil::eic::Client>; 9],
+}
+
+impl<'a> hil::eic::ExternalInterruptController for Eic<'a> {
+    fn enable(&self) {
+        pm::enable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(true);
+        debug!("{}", pm::is_clock_enabled(Clock::PBD(PBDClock::EIC)));
+    }
+
+    fn disable(&self) {
+        pm::disable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(false);
+    }
+
+    fn line_enable(&self, line_num: usize) {
+        let regs: &EicRegisters = &*self.registers;
+        let mask: u32 = 1 << line_num;
```

Comment:
I would prefer doing it as follows to avoid `shift left overflows`  and remove some boiler plate:

```rustc
if line_num <= 8 {
   let interrupt_line = 1 << line_num;
   let regs: &EicRegisters = &*self.registers;
   regs.en.write(Interrupt::INT.val(interrupt_line))
} else {
  debug!("interrupt line: {} not supported!", line_num);
}
``

Reply:
Thanks!